jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -0,0 +1,116 @@
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.ByteBuffer;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public interface Index<V> extends OpaqueIndex
+{
+    @Override
+    default boolean contains(String key)
+    {
+        return get(key) != null;
+    }
+
+    /**
+     * Get an exact match from a String key
+     *
+     * @param s The key
+     * @return the value for the string key
+     */
+    V get(String s);
+
+    /**
+     * Get an exact match from a String key
+     *
+     * @param s The key
+     * @param offset The offset within the string of the key
+     * @param len the length of the key
+     * @return the value for the string / offset / length
+     */
+    V get(String s, int offset, int len);
+
+    /**
+     * Get the best match from key in a byte buffer.
+     * The key is assumed to by ISO_8859_1 characters.
+     *
+     * @param b The buffer
+     * @param offset The offset within the buffer of the key
+     * @param len the length of the key
+     * @return The value or null if not found
+     */
+    V getBest(ByteBuffer b, int offset, int len);
+
+    class Builder<V>
+    {
+        private Map<String, V> contents = new LinkedHashMap<>();
+        private boolean caseSensitive;
+
+        public Builder<V> caseSensitive(boolean caseSensitive)
+        {
+            this.caseSensitive = caseSensitive;
+            return this;
+        }
+
+        public <T> Builder<V> withAllDerivate(T[] values, Function<T, Map<String, V>> function)
+        {
+            for (T value : values)
+            {
+                Map<String, V> pairs = function.apply(value);
+                contents.putAll(pairs);
+            }
+            return this;
+        }
+
+        public Builder<V> withAll(V[] values, Function<V, String> function)
+        {
+            for (V value : values)
+            {
+                String key = function.apply(value);
+                contents.put(key, value);
+            }
+            return this;
+        }
+
+        public Builder<V> with(String key, V value)
+        {
+            contents.put(key, value);
+            return this;
+        }
+
+        public Index<V> build()
+        {
+            Trie<V> result;
+            if (caseSensitive)
+                result = new ArrayTernaryTrie<>(false, contents.size());
+            else
+                result = new ArrayTrie<>(contents.size());
```

Comment:
`contents.size()` is not really the right capacity for ArrayTrie.    Capacity is hard to estimate, as it is more character by character and will be different for `["Foo1","Foo2","Foo3"]` vs `[ "Foo0","Bar0","Bob0"]`.

So I think we should just probably loop here starting at a low capacity, then recreating larger and larger as need be.

Perhaps the builder can have a `maxCapacity(int)` method?